apiserver/admin.go

```diff
@@ -246,58 +259,35 @@ func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
 		result.userLogin = false
 	}
 
-	// Only attempt to login with credentials if we are not doing an anonymous login.
-	var (
-		lastConnection *time.Time
-		// Anonymous logins are for other controllers (for cross-model
-		// relations) - we don't need to start pingers because we
-		// don't maintain presence information for them.
-		startPinger = !result.anonymousLogin
-	)
+	// Anonymous logins come from other controllers (in cross-model relations).
+	// We don't need to start pingers because we don't maintain presence
+	// information for them.
+	startPinger := !result.anonymousLogin
 
+	// Only attempt to login with credentials if we are not doing an anonymous login.
 	if !result.anonymousLogin {
-		// If the user attempted to login to a migrated model,
-		// a.root.model will be nil as the model document does not
-		// exist on this controller and a.root.modelUUID cannot be
-		// resolved. In this case use the requested model UUID to check
-		// if we need to redirect the user.
-		modelUUID := a.root.modelUUID
-		if a.root.model != nil {
-			modelUUID = a.root.model.UUID()
-		}
-
-		if err := a.maybeEmitRedirectError(modelUUID, result.tag); err != nil {
-			return nil, err
-		}
-
-		authInfo, err := a.srv.authenticator.AuthenticateLoginRequest(
-			a.root.serverHost,
-			modelUUID,
-			req,
-		)
+		authInfo, err := a.srv.authenticator.AuthenticateLoginRequest(a.root.serverHost, modelUUID, req)
 		if err != nil {
 			return nil, a.handleAuthError(err)
 		}
+
 		result.controllerMachineLogin = authInfo.Controller
-		// controllerConn is used to indicate a connection from the controller
-		// to a non-controller model.
+		// controllerConn is used to indicate a connection from
+		// the controller to a non-controller model.
 		controllerConn := false
 		if authInfo.Controller && !a.root.state.IsController() {
 			// We only need to run a pinger for controller machine
 			// agents when logging into the controller model.
 			startPinger = false
 			controllerConn = true
 		}
-		a.root.entity = authInfo.Entity
+
 		// TODO(wallyworld) - we can't yet observe anonymous logins as entity must be non-nil
-		a.apiObserver.Login(
-			authInfo.Entity.Tag(),
-			a.root.model.ModelTag(),
-			controllerConn,
-			req.UserData,
-		)
-	} else if a.root.model == nil { // anonymous login to unknown model
-		// Hide the fact that the model does not exist
+		a.root.entity = authInfo.Entity
+		a.apiObserver.Login(authInfo.Entity.Tag(), a.root.model.ModelTag(), controllerConn, req.UserData)
+	} else if a.root.model == nil {
+		// Anonymous login to unknown model.
+		// Hide the fact that the model does not exist.
```

Comment:
We should add login stats here at some point, to at least take note of how many times this happens etc.